### DISCUSSION TITLE
Question: the cucumbers library would also automatically intercept command line parameters, resulting in an unexpected argument '* found error being reported

### DISCUSSION THREAD

**Discussion by konghaiya** [cucumber-rs/cucumber#294] [Q&A] [answered]
## Question:
We used cucumbers to create our testing framework, but we found that when we wanted to add the clip:: Parser startup parameter to the testing framework, the cucumbers library would also automatically intercept command line parameters, resulting in an unexpected argument '* found error being reported
But in the startup parameters added in the upper layer, the - h option can take effect

Is there a way for cucumbers to either block or ignore command-line parameters

**Comment by ilslv** [comment, marked as answer]
You can add custom CLI options, as [shown in the book](https://cucumber-rs.github.io/cucumber/main/cli.html#customizing). Or you can just create a struct of CLI options and pass it via [`Cucumber::with_cli`](https://docs.rs/cucumber/latest/cucumber/struct.Cucumber.html#method.with_cli)

**Reply by konghaiya** [reply]
Okay, thank you. I'll take a look at the relevant interface documents again

**Reply by konghaiya** [reply]
The method you provided does not seem to solve my problem. It seems that using the method you provided can only implement startup parameters based on cucumbers. However, my situation is that I already have startup parameters implemented based on clip:: Parser, and now I need to disable cucumbers' access to startup parameters or ignore them

**Reply by konghaiya** [reply]
Do there currently exist such methods or interfaces

**Reply by ilslv** [reply]
I don't quite understand the problem. Using [`Cucumber::with_cli()`](https://docs.rs/cucumber/latest/cucumber/struct.Cucumber.html#method.with_cli) you can do something like:

```rust
let your_cli = YourCli::parse().unwrap();
let cucumber_cli = cli::Opts {
    re_filter: None,
    tags_filter: None,
    parser: your_cli.parser,
    runner:your_cli.runner,
    writer: your_cli.writer,
    custom: your_cli.custom,
}
YourWorld::cucumber()
    .with_cli(cucumber_cli)
    .run_and_exit("tests/features/readme")
    .await;
```

In this case stdin won't be touched by cucumber at all.

**Reply by konghaiya** [reply]
```
#[derive(Parser, Debug, Default)]
#[command(author, version, about, long_about = None)]
pub struct ClientArgs {
    #[arg(short, long)]
    /// Login User Name
    pub user: Option<String>,

    #[arg(short, long)]
    /// Login User Password
    pub password: Option<String>,

    #[arg(short, long)]
    /// The IP address and port of the service node that needs to be logged in
    pub endpoint: Option<String>,

    #[arg(long)]
    /// The maximum number of rows that the client reads data from the server at a time
    pub batch_size: Option<i64>,

    #[arg(long)]
    /// in seconds
    pub connect_timeout: Option<u32>,

    #[arg(long)]
    /// The specified configuration file path
    pub config_path: Option<String>,

    #[arg(short, long)]
    /// query text
    pub input: Option<String>,

    #[arg(short, long)]
    /// query text file
    pub file: Option<String>,
}
```

**Reply by konghaiya** [reply]
Thank you very much for your answer. 
The above is the command line structure I defined, and there are no members like Parser, Runner, etc. How can I modify it?
 Looking forward to your reply

**Reply by ilslv** [reply]
Components like `Parser`, `Runner` and `Writer` can be swapped, so traits defining them have associated types like [`Parser::Cli`](https://docs.rs/cucumber/latest/cucumber/trait.Parser.html#associatedtype.Cli). By default it's [`parser::basic::Cli`](https://docs.rs/cucumber/latest/cucumber/parser/basic/struct.Cli.html). Same for [`runner::basic::Cli`](https://docs.rs/cucumber/latest/cucumber/runner/basic/struct.Cli.html) and [`writer::basic::Cli`](https://docs.rs/cucumber/latest/cucumber/writer/basic/struct.Cli.html). You can fill them by hand with constants or lift some of the arguments to the `ClientArgs` and pass them from there. 

All of that is covered in the [CLI book chapter](https://cucumber-rs.github.io/cucumber/main/cli.html), please read it.

**Reply by konghaiya** [reply]
Okay, thank you. I'll take a look at the relevant interface documents again

**Reply by konghaiya** [reply]
Thank you very much. My problem has been resolved
Use the following method to register custom startup parameters into the cucumber
// Register startup parameters into the cucumber framework
    let opts = cli::Opts::<_, _, _, ClientArgs>::parsed();

 queryWorld::cucumber().with_cli(opts);

**Comment by konghaiya** [comment]
@ilslv 
Sorry to bother you, but I noticed that after doing so, I will block the parameters of the cucumber itself. Is there a way to make both the custom parameters and the cucumber parameters take effect

**Reply by konghaiya** [reply]
For some reason, we must use custom command line parameters, but at the same time, we hope that it does not affect the parameter function of the cucumber itself. I apologize for my limited ability and cannot find a solution in the documentation you provided. I hope you can provide me with an answer. Thank you very much
